pennylane/devices/capabilities.py

```diff
@@ -0,0 +1,328 @@
+    "mid_circuit_measurements": False,
+    "runtime_code_generation": False,
+    "dynamic_qubit_management": False,
+    "overlapping_observables": True,
+    "non_commuting_observables": False,
+    "initial_state_prep": False,
+}
+
+
+def _get_toml_section(document: dict, path: str, prefix: str = "") -> dict:
+    """Retrieves a section from a TOML document using a given path."""
+
+    if prefix:
+        path = f"{prefix}.{path}"
+
+    for k in path.split("."):
+        if not isinstance(document, dict) or k not in document:
+            return {}
+        document = document[k]
+    return document
+
+
+def _validate_conditions(conditions: list[ExecutionCondition], target=None) -> None:
+    """Validates the execution conditions."""
+
+    if (
+        ExecutionCondition.ANALYTIC_MODE_ONLY in conditions
+        and ExecutionCondition.FINITE_SHOTS_ONLY in conditions
+    ):
+        raise InvalidCapabilitiesError(
+            "Conditions cannot contain both 'analytic' and 'finiteshots'"
+        )
+
+    if ExecutionCondition.TERMS_MUST_COMMUTE in conditions and target not in (
+        "Prod",
+        "SProd",
+        "Sum",
+        "LinearCombination",
+        "Hamiltonian",
+    ):
+        raise InvalidCapabilitiesError(
+            "'terms-commute' is only applicable to Prod, SProd, Sum, and LinearCombination."
+        )
+
+
+def _get_operators(section: dict) -> dict[str, OperatorProperties]:
+    """Parses an operator section into a dictionary of OperatorProperties."""
+
+    operators = {}
+    iterator = section.items() if hasattr(section, "items") else zip(section, repeat({}))
+    for o, attributes in iterator:
+        if unknowns := set(attributes) - {"properties", "conditions"}:
+            raise InvalidCapabilitiesError(
+                f"Operator '{o}' has unknown attributes: {list(unknowns)}"
+            )
+        properties = attributes.get("properties", {})
+        if unknowns := set(properties) - {"invertible", "controllable", "differentiable"}:
+            raise InvalidCapabilitiesError(
+                f"Operator '{o}' has unknown properties: {list(unknowns)}"
+            )
+        condition_strs = attributes.get("conditions", [])
+        if unknowns := set(condition_strs) - VALID_CONDITION_STRINGS:
+            raise InvalidCapabilitiesError(
+                f"Operator '{o}' has unknown conditions: {list(unknowns)}"
+            )
+        conditions = [ExecutionCondition(c) for c in condition_strs]
+        _validate_conditions(conditions, o)
+        operators[o] = OperatorProperties(
+            invertible="invertible" in properties,
+            controllable="controllable" in properties,
+            differentiable="differentiable" in properties,
+            conditions=conditions,
+        )
+    return operators
+
+
+def _get_operations(document: dict, prefix: str = "") -> dict[str, OperatorProperties]:
+    """Gets the supported operations from a TOML document."""
```

Comment:
it took me a while to figure out what the prefix was for. Currently the docstring only mentions the TOML document, and we include the `prefix` in the docstrings as well?